#### 1.	What is a neural network? What are the general steps required to build a neural network? 

## Neural Networks

### General idea 
The NN is a algorithm that mimics the human brain to understand patterns in data by creating a series of models and algorithms to do so. 

### Definition
A Neural network attempts to classify the target variable accurately by determining which input is most helpful in classifying data without error.

### Structure of a Neural Network
Composed of layers that are made up of nodes. 
    
    Node: place where computation happens--"perceptron" 

- Input Layer = is the input data

- Output Layer = is the prediction

- Hidden Layers: Much of the work and processes take place in the hidden layers. Its not something that is directly observable. 

### How Neural Networks Work

1. A node takes in multiple variables (features) as input from the data and combines it with a set of coefficients or weights (for the purpose of making predictions) 

This particular process can do one of the following: 

    - Amplify input
    - Dampen input
      
2. The input-weight combnations are summed

3. The sum is then passed through the activation function which determines whether or not that "signal" should continue through the network. If it does continue, the activation function determines the extent to which the signal progresses through the neural network. Once the signal passes, the node is “activated.” NOTE: Multiple linear regression (more than one feature affecting the outcome) is happening at every node of a neural network.

- Then each node is compared with other nodes iteratively

4. Eventually this process leads to the ultimate outcome. 


Each layer’s output is simultaneously the subsequent layer’s input, starting from an initial input layer receiving your data.

#### 2.	Generally, how do you check the performance of a neural network? Why? 

## Loss Function

Loss Functions are used to check the performance of a neural network. 

### What is a Loss Function?

- This is a function that calculates the error of a given set of weights in order to determine the distance between the neural network's current or predicted output and the expected output ot target value. 

### How is the 'loss' from a Loss function used?

1) The loss is calculated

        - If the deviation between y_pred(predicted output) and y(expected output/target value) is very large,
          the loss value will be very high.
        - If the deviation is small or the values are nearly identical, loss value will be low.
    
2) LEARNING PROCESS: the loss is used as a feedback signal to adjust the way in which the Nework works


### Types of Loss Functions

There are different types of loss functions to choose from to calculate the loss of the model. But it is important to choose the right one. We have different loss functions because of the different types of models out there--like regression, classification models. This difference is also reflected in the types of loss functions. 

PyTorch’s torch.nn module contains the types of loss functions. 


### Why the need to check performance?

We need to see if our model is doing a good job in predicting our target variable. 

#### 3.	Create a neural network using keras to predict the outcome of either of these datasets: 
Cardiac Arrhythmia: https://archive.ics.uci.edu/ml/datasets/Arrhythmia 
Abalone age: https://archive.ics.uci.edu/ml/datasets/Abalone

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn import tree
from sklearn.metrics import classification_report, plot_confusion_matrix
import pydotplus
from IPython.display import Image


from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

In [305]:
arrhythmia_df = pd.read_csv("arrhythmia.data") 
arrhythmia_df

age  sex  height  weight  QRSduration  PRinterval  Q-Tinterval  \
0     75    0     190      80           91         193          371   
1     56    1     165      64           81         174          401   
2     54    0     172      95          138         163          386   
3     55    0     175      94          100         202          380   
4     75    0     190      80           88         181          360   
..   ...  ...     ...     ...          ...         ...          ...   
447   53    1     160      70           80         199          382   
448   37    0     190      85          100         137          361   
449   36    0     166      68          108         176          365   
450   32    1     155      55           93         106          386   
451   78    1     160      70           79         127          364   

     Tinterval  Pinterval  QRS  ... chV6_QwaveAmp chV6_RwaveAmp chV6_SwaveAmp  \
0          174        121  -16  ...           0.0           9.0          -0.9   
1          149         39   25  ...           0.0           8.5           0.0   
2          185        102   96  ...           0.0           9.5          -2.4   
3          179        143   28  ...           0.0          12.2          -2.2   
4          177        103  -16  ...           0.0          13.1          -3.6   
..         ...        ...  ...  ...           ...           ...           ...   
447        154        117  -37  ...           0.0           4.3          -5.0   
448        201         73   86  ...           0.0          15.6          -1.6   
449        194        116  -85  ...           0.0          16.3         -28.6   
450        218         63   54  ...          -0.4          12.0          -0.7   
451        138         78   28  ...           0.0          10.4          -1.8   

    chV6_RPwaveAmp chV6_SPwaveAmp  chV6_PwaveAmp  chV6_TwaveAmp  chV6_QRSA  \
0              0.0            0.0            0.9            2.9       23.3   
1              0.0            0.0            0.2            2.1       20.4   
2              0.0            0.0            0.3            3.4       12.3   
3              0.0            0.0            0.4            2.6       34.6   
4              0.0            0.0           -0.1            3.9       25.4   
..             ...            ...            ...            ...        ...   
447            0.0            0.0            0.7            0.6       -4.4   
448            0.0            0.0            0.4            2.4       38.0   
449            0.0            0.0            1.5            1.0      -44.2   
450            0.0            0.0            0.5            2.4       25.0   
451            0.0            0.0            0.5            1.6       21.3   

     chV6_QRSTA  class  
0          49.4      8  
1          38.8      6  
2          49.0     10  
3          61.6      1  
4          62.8      7  
..          ...    ...  
447        -0.5      1  
448        62.4     10  
449       -33.2      2  
450        46.6      1  
451        32.8      1  

[452 rows x 280 columns]

## 1) There are no nulls in the dataset!

## 2) Two rows are removed below...

### (In other words, 450 rows are "pulled" into the variable arrhythmia_df, heights --> 22 & 23ft)

In [306]:
arrhythmia_df = arrhythmia_df.loc[arrhythmia_df['height'] < 274]

## 3) '?' Replaced with 'NaN'...

In [307]:
arrhythmia_df = arrhythmia_df.replace('?', np.nan)

In [308]:
for i in arrhythmia_df.isnull().sum():
    if i > 0:
        print(i)

8
22
1
374
1


In [309]:
print("T", (arrhythmia_df["T"].isnull().sum()))
print("P", (arrhythmia_df["P"].isnull().sum()))
print("QRST", (arrhythmia_df["QRST"].isnull().sum()))
print("J", (arrhythmia_df["J"].isnull().sum()))
print("heartrate", (arrhythmia_df["heartrate"].isnull().sum()))

T 8
P 22
QRST 1
J 374
heartrate 1


## 4) Dropping Columns with Majority NaNs...

### The J Column (TOO MANY 'NaNs')

In [310]:
#83% are nulls--this column will be dropped
arrhythmia_df = arrhythmia_df.drop(['J'], axis = 1)

### THIS IS ONE OF THE ERRORS I FOUND! (FIXED!)

#### NOTE: .isnull.sum( ) is not recognizing "NaN" values 

EX: arrhythmia_df["heartrate"].isnull().sum()--> There is 1 NaN value in the dataset!

In [28]:
#This was ran BEFORE fixing the error('?' was not properly replaced by NaN)
arrhythmia_df["heartrate"].isnull().sum()

0

In [311]:
arrhythmia_df

age  sex  height  weight  QRSduration  PRinterval  Q-Tinterval  \
0     75    0     190      80           91         193          371   
1     56    1     165      64           81         174          401   
2     54    0     172      95          138         163          386   
3     55    0     175      94          100         202          380   
4     75    0     190      80           88         181          360   
..   ...  ...     ...     ...          ...         ...          ...   
447   53    1     160      70           80         199          382   
448   37    0     190      85          100         137          361   
449   36    0     166      68          108         176          365   
450   32    1     155      55           93         106          386   
451   78    1     160      70           79         127          364   

     Tinterval  Pinterval  QRS  ... chV6_QwaveAmp chV6_RwaveAmp chV6_SwaveAmp  \
0          174        121  -16  ...           0.0           9.0          -0.9   
1          149         39   25  ...           0.0           8.5           0.0   
2          185        102   96  ...           0.0           9.5          -2.4   
3          179        143   28  ...           0.0          12.2          -2.2   
4          177        103  -16  ...           0.0          13.1          -3.6   
..         ...        ...  ...  ...           ...           ...           ...   
447        154        117  -37  ...           0.0           4.3          -5.0   
448        201         73   86  ...           0.0          15.6          -1.6   
449        194        116  -85  ...           0.0          16.3         -28.6   
450        218         63   54  ...          -0.4          12.0          -0.7   
451        138         78   28  ...           0.0          10.4          -1.8   

    chV6_RPwaveAmp  chV6_SPwaveAmp  chV6_PwaveAmp  chV6_TwaveAmp  chV6_QRSA  \
0              0.0             0.0            0.9            2.9       23.3   
1              0.0             0.0            0.2            2.1       20.4   
2              0.0             0.0            0.3            3.4       12.3   
3              0.0             0.0            0.4            2.6       34.6   
4              0.0             0.0           -0.1            3.9       25.4   
..             ...             ...            ...            ...        ...   
447            0.0             0.0            0.7            0.6       -4.4   
448            0.0             0.0            0.4            2.4       38.0   
449            0.0             0.0            1.5            1.0      -44.2   
450            0.0             0.0            0.5            2.4       25.0   
451            0.0             0.0            0.5            1.6       21.3   

     chV6_QRSTA  class  
0          49.4      8  
1          38.8      6  
2          49.0     10  
3          61.6      1  
4          62.8      7  
..          ...    ...  
447        -0.5      1  
448        62.4     10  
449       -33.2      2  
450        46.6      1  
451        32.8      1  

[450 rows x 279 columns]

### CURRENT COUNT: (279, 450)

In [312]:
type(arrhythmia_df)

pandas.core.frame.DataFrame

## 4) Interpolating... (Didn't Work)


### (use simple imputer from sklearn)

In [165]:
#arrhythmia_df = arrhythmia_df.interpolate(method='polynomial', order=1, axis=0, limit_area='inside')

### Confirming Interpolation... (Didn't Work)

In [166]:
for i in arrhythmia_df.isnull().sum():
    if i > 0:
        print(i)

8
22
1
1


In [283]:
print("T", (arrhythmia_df["T"].isnull().sum()))
print("P", (arrhythmia_df["P"].isnull().sum()))
print("QRST", (arrhythmia_df["QRST"].isnull().sum()))
print("heartrate", (arrhythmia_df["heartrate"].isnull().sum()))

T 8
P 22
QRST 1
heartrate 1


## 5) Trying to replace 'NaNs' with Mode or Mean of Column...

In [169]:
#arrhythmia_df[entered_all_column_names_above].value_counts()

### NOTE: Look into how to code the mean for dataset containing negative & positive number. 

#### (In the meantime, I used mode to replace NaNs)

In [313]:
#All Columns to a list 
All_Columns = arrhythmia_df.columns.tolist()

In [314]:
NaNs_list = []
for i in All_Columns:
    null_amount = arrhythmia_df[i].isnull().sum()
    if null_amount > 0:
        NaNs_list.append(i)

In [315]:
NaNs_list

['T', 'P', 'QRST', 'heartrate']

In [316]:
for i in NaNs_list:
    arrhythmia_df[i].fillna(arrhythmia_df[i].mode()[0], inplace=True)

### Confirming NaNs are Filled...

In [317]:
print("T", (arrhythmia_df["T"].isnull().sum()))
print("P", (arrhythmia_df["P"].isnull().sum()))
print("QRST", (arrhythmia_df["QRST"].isnull().sum()))
print("heartrate", (arrhythmia_df["heartrate"].isnull().sum()))

T 0
P 0
QRST 0
heartrate 0


In [318]:
type(arrhythmia_df)

pandas.core.frame.DataFrame

## 6) Transform Categorical Data

#### Do all columns need to have the same datatype?

In [319]:
for col in arrhythmia_df:
    if (arrhythmia_df[col].dtypes) == object:
        print(col, arrhythmia_df[col].dtypes)

T object
P object
QRST object
heartrate object


In [320]:
for col in arrhythmia_df:
    if (arrhythmia_df[col].dtypes) == object:
        arrhythmia_df[col] = arrhythmia_df[col].astype("int")

In [321]:
for col in arrhythmia_df:
    print(arrhythmia_df[col].dtypes)

int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
float64
float64
float64
float64
float64
float64
floa

### TRY SKIPPING THIS STEPS 7-8 IF MODEL GIVES ERROR...something is fishy here...where did the NaNs come from?

## 7) Checking the Variance...

As you can see below there is alot of variance between the columns!

In [179]:
#for col in arrhythmia_df:
    #print(col, arrhythmia_df[col].var())

age 262.9011828755259
sex 0.24793862905221478
height 108.41380351398169
weight 260.26449393714427
QRSduration 237.00333580796834
PRinterval 2017.676594902252
Q-Tinterval 1040.727389260084
Tinterval 1270.5470378619154
Pinterval 667.4107597129423
QRS 2047.9572828507792
T 3306.6154120267256
P 805.6373224449393
QRST 1290.6829200692898
heartrate 166.53563474387528
chDI_Qwave 113.78601336302894
chDI_Rwave 331.0559564464242
chDI_Swave 422.0915614946795
chDI_RPwave 2.474159861420441
chDI_SPwave 0.0
chDI_intrinsicReflecttions 100.2113140311804
chDI_RRwaveExists 0.0022222222222222227
chDI_DD_RRwaveExists 0.011012125711457564
chDI_RPwaveExists 0.011012125711457564
chDI_DD_RPwaveExists 0.0044345459044790895
chDI_RTwaveExists 0.0044345459044790895
chDI_DD_RTwaveExists 0.008829497649096758
chDII_Qwave 126.32318732986883
chDII_Rwave 297.2504627567434
chDII_Swave 445.25699579312055
chDII_RPwave 8.75537738183618
chDII_SPwave 6.96888888888889
chDII_intrinsicReflecttions 92.44642415243752
chDII_RRwaveExi

In [180]:
#normalizing dataframe
#arrhythmia_df = arrhythmia_df.apply(lambda iterator: ((iterator.max() - iterator)/(iterator.max() - iterator.min())).round(2)) 

In [181]:
#for col in arrhythmia_df:
    #print(col, arrhythmia_df[col].var())

age 0.039104516703786195
sex 0.24793862905221478
height 0.014917698094531055
weight 0.009421692155407077
QRSduration 0.01335203365503588
PRinterval 0.007346657263053699
Q-Tinterval 0.014395203662459789
Tinterval 0.017010599356594904
Pinterval 0.01593821380846325
QRS 0.01760853699579312
T 0.02608423113090819
P 0.006720251917842118
QRST 0.014031539222964611
heartrate 0.02600395941598614
chDI_Qwave 0.014710548874041078
chDI_Rwave 0.013509202672605788
chDI_Swave 0.054486414253897546
chDI_RPwave 0.00430525167037862
chDI_SPwave nan
chDI_intrinsicReflecttions 0.01002113140311804
chDI_RRwaveExists 0.0022222222222222227
chDI_DD_RRwaveExists 0.011012125711457564
chDI_RPwaveExists 0.011012125711457564
chDI_DD_RPwaveExists 0.00443454590447909
chDI_RTwaveExists 0.00443454590447909
chDI_DD_RTwaveExists 0.008829497649096758
chDII_Qwave 0.021894895322939863
chDII_Rwave 0.017587639198218263
chDII_Swave 0.05256561494679534
chDII_RPwave 0.006771853006681518
chDII_SPwave 0.0022222222222222227
chDII_intrin

## 8) Remove Columns with Majority Nulls--> Resulting from Normalization

## (step 7 not ran, no need for step 8)
## (no NaNs in the dataset)

In [262]:
list_col = []
for col in arrhythmia_df:
    if (arrhythmia_df[col].isnull().sum()) > 0:
        list_col.append(col)
        print(col, (arrhythmia_df[col].isnull().sum()))

In [263]:
list_col

[]

In [184]:
#arrhythmia_df = arrhythmia_df.drop(['chDI_SPwave','chAVL_SPwave','chAVL_RRwaveExists',
                                              #'chAVL_DD_RTwaveExists','chAVF_RPwaveExists',
                                              #'chV4_RPwaveExists','chV4_DD_RPwaveExists','chV5_SPwave',
                                              #'chV5_RRwaveExists', 'chV5_RPwaveExists','chV5_RTwaveExists',
                                              #'chV6_SPwave','chV6_DD_RPwaveExists','chV6_RTwaveExists',
                                              #'chDI_SPwaveAmp','chAVL_SPwaveAmp','chV5_SPwaveAmp',
                                              #'chV6_SPwaveAmp'], axis = 1)

### COUNT:  261, 450 (IF STEPS 7-8 WERE USED)

### CURRENT VALUE : 279, 450 

In [322]:
arrhythmia_df

age  sex  height  weight  QRSduration  PRinterval  Q-Tinterval  \
0     75    0     190      80           91         193          371   
1     56    1     165      64           81         174          401   
2     54    0     172      95          138         163          386   
3     55    0     175      94          100         202          380   
4     75    0     190      80           88         181          360   
..   ...  ...     ...     ...          ...         ...          ...   
447   53    1     160      70           80         199          382   
448   37    0     190      85          100         137          361   
449   36    0     166      68          108         176          365   
450   32    1     155      55           93         106          386   
451   78    1     160      70           79         127          364   

     Tinterval  Pinterval  QRS  ...  chV6_QwaveAmp  chV6_RwaveAmp  \
0          174        121  -16  ...            0.0            9.0   
1          149         39   25  ...            0.0            8.5   
2          185        102   96  ...            0.0            9.5   
3          179        143   28  ...            0.0           12.2   
4          177        103  -16  ...            0.0           13.1   
..         ...        ...  ...  ...            ...            ...   
447        154        117  -37  ...            0.0            4.3   
448        201         73   86  ...            0.0           15.6   
449        194        116  -85  ...            0.0           16.3   
450        218         63   54  ...           -0.4           12.0   
451        138         78   28  ...            0.0           10.4   

     chV6_SwaveAmp  chV6_RPwaveAmp  chV6_SPwaveAmp  chV6_PwaveAmp  \
0             -0.9             0.0             0.0            0.9   
1              0.0             0.0             0.0            0.2   
2             -2.4             0.0             0.0            0.3   
3             -2.2             0.0             0.0            0.4   
4             -3.6             0.0             0.0           -0.1   
..             ...             ...             ...            ...   
447           -5.0             0.0             0.0            0.7   
448           -1.6             0.0             0.0            0.4   
449          -28.6             0.0             0.0            1.5   
450           -0.7             0.0             0.0            0.5   
451           -1.8             0.0             0.0            0.5   

     chV6_TwaveAmp  chV6_QRSA  chV6_QRSTA  class  
0              2.9       23.3        49.4      8  
1              2.1       20.4        38.8      6  
2              3.4       12.3        49.0     10  
3              2.6       34.6        61.6      1  
4              3.9       25.4        62.8      7  
..             ...        ...         ...    ...  
447            0.6       -4.4        -0.5      1  
448            2.4       38.0        62.4     10  
449            1.0      -44.2       -33.2      2  
450            2.4       25.0        46.6      1  
451            1.6       21.3        32.8      1  

[450 rows x 279 columns]

#### REMINDER: 1.00 = Regular Heartbeat

In [186]:
arrhythmia_df['class'].value_counts()

1.00    245
0.40     50
0.93     44
0.67     25
0.00     22
0.87     15
0.80     15
0.73     11
0.47      9
0.07      5
0.13      4
0.60      3
0.53      2
Name: class, dtype: int64

In [265]:
type(arrhythmia_df)

pandas.core.frame.DataFrame

## 9) Dropping Columns through Correlation Matrix

In [323]:
cor_matrix = arrhythmia_df.corr().abs() #abs: takes absolute value of correlation values

In [324]:
from sklearn import datasets

upper_tri = cor_matrix.where(np.triu(np.ones(cor_matrix.shape),k=1).astype(bool))
upper_tri

age       sex    height    weight  QRSduration  PRinterval  \
age            NaN  0.061665  0.243302  0.352977     0.008168    0.039119   
sex            NaN       NaN  0.460879  0.257447     0.338382    0.048942   
height         NaN       NaN       NaN  0.574054     0.047447    0.081453   
weight         NaN       NaN       NaN       NaN     0.097892    0.120221   
QRSduration    NaN       NaN       NaN       NaN          NaN    0.021415   
...            ...       ...       ...       ...          ...         ...   
chV6_PwaveAmp  NaN       NaN       NaN       NaN          NaN         NaN   
chV6_TwaveAmp  NaN       NaN       NaN       NaN          NaN         NaN   
chV6_QRSA      NaN       NaN       NaN       NaN          NaN         NaN   
chV6_QRSTA     NaN       NaN       NaN       NaN          NaN         NaN   
class          NaN       NaN       NaN       NaN          NaN         NaN   

               Q-Tinterval  Tinterval  Pinterval       QRS  ...  \
age               0.154391   0.015105   0.098167  0.251801  ...   
sex               0.072672   0.187283   0.085497  0.072584  ...   
height            0.049501   0.047882   0.132786  0.128398  ...   
weight            0.058241   0.140564   0.120577  0.153145  ...   
QRSduration       0.220961   0.396953   0.049144  0.144556  ...   
...                    ...        ...        ...       ...  ...   
chV6_PwaveAmp          NaN        NaN        NaN       NaN  ...   
chV6_TwaveAmp          NaN        NaN        NaN       NaN  ...   
chV6_QRSA              NaN        NaN        NaN       NaN  ...   
chV6_QRSTA             NaN        NaN        NaN       NaN  ...   
class                  NaN        NaN        NaN       NaN  ...   

               chV6_QwaveAmp  chV6_RwaveAmp  chV6_SwaveAmp  chV6_RPwaveAmp  \
age                 0.166185       0.177108       0.061554        0.091021   
sex                 0.228771       0.043770       0.091940        0.027427   
height              0.061722       0.014724       0.071417        0.008875   
weight              0.050264       0.048485       0.023966        0.047582   
QRSduration         0.199545       0.089928       0.233070        0.013813   
...                      ...            ...            ...             ...   
chV6_PwaveAmp            NaN            NaN            NaN             NaN   
chV6_TwaveAmp            NaN            NaN            NaN             NaN   
chV6_QRSA                NaN            NaN            NaN             NaN   
chV6_QRSTA               NaN            NaN            NaN             NaN   
class                    NaN            NaN            NaN             NaN   

               chV6_SPwaveAmp  chV6_PwaveAmp  chV6_TwaveAmp  chV6_QRSA  \
age                       NaN       0.024823       0.273189   0.004119   
sex                       NaN       0.009666       0.068842   0.028777   
height                    NaN       0.051245       0.067138   0.057516   
weight                    NaN       0.028774       0.145250   0.035322   
QRSduration               NaN       0.064731       0.221894   0.128127   
...                       ...            ...            ...        ...   
chV6_PwaveAmp             NaN            NaN       0.041673   0.064723   
chV6_TwaveAmp             NaN            NaN            NaN   0.121468   
chV6_QRSA                 NaN            NaN            NaN        NaN   
chV6_QRSTA                NaN            NaN            NaN        NaN   
class                     NaN            NaN            NaN        NaN   

               chV6_QRSTA     class  
age              0.221985  0.090826  
sex              0.045815  0.178393  
height           0.017255  0.033703  
weight           0.075807  0.088700  
QRSduration      0.085211  0.324365  
...                   ...       ...  
chV6_PwaveAmp    0.058376  0.088195  
chV6_TwaveAmp    0.693580  0.031134  
chV6_QRSA        0.607237  0.005520  
chV6_QRSTA            NaN  0.007727  
class                 NaN       NaN  

[2

In [325]:
to_drop = [column for column in upper_tri.columns if any(upper_tri[column] > 0.75)]
print(to_drop)

['chAVR_Swave', 'chAVR_intrinsicReflecttions', 'chAVR_DD_RTwaveExists', 'chAVF_Swave', 'chV1_Swave', 'chV2_Rwave', 'chV3_intrinsicReflecttions', 'chV6_Qwave', 'chV6_Rwave', 'chV6_intrinsicReflecttions', 'chDI_QwaveAmp', 'chDI_RPwaveAmp', 'chDI_QRSA', 'chDII_QwaveAmp', 'chDII_SwaveAmp', 'chDII_RPwaveAmp', 'chDII_SPwaveAmp', 'chDII_QRSA', 'chDIII_QwaveAmp', 'chDIII_RwaveAmp', 'chDIII_SwaveAmp', 'chDIII_RPwaveAmp', 'chDIII_SPwaveAmp', 'chDIII_PwaveAmp', 'chDIII_QRSA', 'chDIII_QRSTA', 'chAVR_JJwaveAmp', 'chAVR_SwaveAmp', 'chAVR_RPwaveAmp', 'chAVR_SPwaveAmp', 'chAVR_PwaveAmp', 'chAVR_TwaveAmp', 'chAVR_QRSTA', 'chAVL_JJwaveAmp', 'chAVL_QwaveAmp', 'chAVL_RwaveAmp', 'chAVL_SwaveAmp', 'chAVL_RPwaveAmp', 'chAVL_TwaveAmp', 'chAVL_QRSA', 'chAVL_QRSTA', 'chAVF_QwaveAmp', 'chAVF_RwaveAmp', 'chAVF_SwaveAmp', 'chAVF_RPwaveAmp', 'chAVF_SPwaveAmp', 'chAVF_PwaveAmp', 'chAVF_TwaveAmp', 'chAVF_QRSA', 'chAVF_QRSTA', 'chV1_QwaveAmp', 'chV1_RPwaveAmp', 'chV1_SPwaveAmp', 'chV2_QwaveAmp', 'chV2_RPwaveAmp', 'chV

In [326]:
arrhythmia_NEW_df = arrhythmia_df.drop(['chAVR_Swave', 'chAVR_intrinsicReflecttions', 
                                         'chAVR_DD_RTwaveExists', 'chAVF_Swave', 'chV1_Swave', 'chV2_Rwave', 
                                         'chV3_intrinsicReflecttions', 'chV6_Qwave', 'chV6_Rwave', 
                                         'chV6_intrinsicReflecttions', 'chDI_QwaveAmp', 'chDI_RPwaveAmp', 
                                         'chDI_QRSA', 'chDII_QwaveAmp', 'chDII_SwaveAmp', 'chDII_RPwaveAmp', 
                                         'chDII_SPwaveAmp', 'chDII_QRSA', 'chDIII_QwaveAmp', 'chDIII_RwaveAmp', 
                                         'chDIII_SwaveAmp', 'chDIII_RPwaveAmp', 'chDIII_SPwaveAmp', 'chDIII_PwaveAmp', 
                                         'chDIII_QRSA', 'chDIII_QRSTA', 'chAVR_JJwaveAmp', 'chAVR_SwaveAmp', 
                                         'chAVR_RPwaveAmp', 'chAVR_SPwaveAmp', 'chAVR_PwaveAmp', 'chAVR_TwaveAmp', 
                                         'chAVR_QRSTA', 'chAVL_JJwaveAmp', 'chAVL_QwaveAmp', 'chAVL_RwaveAmp', 
                                         'chAVL_SwaveAmp', 'chAVL_RPwaveAmp', 'chAVL_TwaveAmp', 'chAVL_QRSA', 
                                         'chAVL_QRSTA', 'chAVF_QwaveAmp', 'chAVF_RwaveAmp', 'chAVF_SwaveAmp', 
                                         'chAVF_RPwaveAmp', 'chAVF_SPwaveAmp', 'chAVF_PwaveAmp', 'chAVF_TwaveAmp', 
                                         'chAVF_QRSA', 'chAVF_QRSTA', 'chV1_QwaveAmp', 'chV1_RPwaveAmp', 
                                         'chV1_SPwaveAmp', 'chV2_QwaveAmp', 'chV2_RPwaveAmp', 'chV2_SPwaveAmp', 
                                         'chV3_JJwaveAmp', 'chV3_QwaveAmp', 'chV3_RwaveAmp', 'chV3_RPwaveAmp', 
                                         'chV3_SPwaveAmp', 'chV3_TwaveAmp', 'chV3_QRSA', 'chV3_QRSTA', 'chV4_JJwaveAmp', 
                                         'chV4_QwaveAmp', 'chV4_SwaveAmp', 'chV4_RPwaveAmp', 'chV4_SPwaveAmp', 
                                         'chV4_TwaveAmp', 'chV5_QwaveAmp', 'chV5_RPwaveAmp', 'chV5_PwaveAmp', 
                                         'chV5_TwaveAmp', 'chV5_QRSTA', 'chV6_JJwaveAmp', 'chV6_QwaveAmp',
                                         'chV6_RwaveAmp', 'chV6_SwaveAmp', 'chV6_RPwaveAmp', 'chV6_PwaveAmp',
                                         'chV6_TwaveAmp', 'chV6_QRSA', 'chV6_QRSTA'], axis = 1)

###  COUNT: 177, 450 (IF STEPS 7-8 WERE USED)

### CURRENT COUNT: 195, 450

In [327]:
arrhythmia_NEW_df

age  sex  height  weight  QRSduration  PRinterval  Q-Tinterval  \
0     75    0     190      80           91         193          371   
1     56    1     165      64           81         174          401   
2     54    0     172      95          138         163          386   
3     55    0     175      94          100         202          380   
4     75    0     190      80           88         181          360   
..   ...  ...     ...     ...          ...         ...          ...   
447   53    1     160      70           80         199          382   
448   37    0     190      85          100         137          361   
449   36    0     166      68          108         176          365   
450   32    1     155      55           93         106          386   
451   78    1     160      70           79         127          364   

     Tinterval  Pinterval  QRS  ...  chV4_PwaveAmp  chV4_QRSA  chV4_QRSTA  \
0          174        121  -16  ...            0.9       17.7        70.7   
1          149         39   25  ...            0.5       11.8        34.6   
2          185        102   96  ...            0.5       -3.0        20.7   
3          179        143   28  ...            0.1       28.8        63.1   
4          177        103  -16  ...           -0.1       16.2        63.2   
..         ...        ...  ...  ...            ...        ...         ...   
447        154        117  -37  ...            0.8      -20.1        -9.5   
448        201         73   86  ...            0.7       69.2       129.3   
449        194        116  -85  ...            1.0      -71.2      -161.4   
450        218         63   54  ...            0.6       17.5        56.2   
451        138         78   28  ...            0.5       44.0        60.8   

     chV5_JJwaveAmp  chV5_RwaveAmp  chV5_SwaveAmp  chV5_SPwaveAmp  chV5_QRSA  \
0              -0.4           13.5           -4.0             0.0       25.5   
1              -0.4           11.0           -2.4             0.0       21.6   
2               1.3           11.1           -3.4             0.0       11.5   
3               0.1           15.2           -3.7             0.0       36.8   
4              -0.2            9.1           -0.9             0.0       21.7   
..              ...            ...            ...             ...        ...   
447             0.1            4.1           -8.3             0.0       -8.4   
448            -0.7           21.2           -2.8             0.0       50.7   
449             0.2           22.0          -30.8             0.0      -39.6   
450             0.1           15.3           -3.5             0.0       29.7   
451            -0.4           12.5           -2.9             0.0       26.7   

     chV6_SPwaveAmp  class  
0               0.0      8  
1               0.0      6  
2               0.0     10  
3               0.0      1  
4               0.0      7  
..              ...    ...  
447             0.0      1  
448             0.0     10  
449             0.0      2  
450             0.0      1  
451             0.0      1  

[450 rows x 195 columns]

## 10) Making Target Binary

### Predictions need to sum to 1 so that they can be interpreted as probability by the NN model!

In [328]:
arrhythmia_NEW_df['class'].value_counts()

1     245
10     50
2      44
6      25
16     22
3      15
4      15
5      11
9       9
15      5
14      4
7       3
8       2
Name: class, dtype: int64

### STEP 1: Replace Regular Heartbeat with '11' as a temporary placeholder

In [329]:
for n in arrhythmia_NEW_df['class']:
    if n == 1:
        arrhythmia_NEW_df = arrhythmia_NEW_df.replace([n], 11)

In [330]:
arrhythmia_NEW_df['class'].value_counts()

11    245
10     50
2      44
6      25
16     22
3      15
4      15
5      11
9       9
15      5
14      4
7       3
8       2
Name: class, dtype: int64

### STEP 2: Replace Irregular Heartbeat with '1' as a permanent "probability" value

In [331]:
for n in arrhythmia_NEW_df['class']:
    if n != 11:
        arrhythmia_NEW_df = arrhythmia_NEW_df.replace([n], 1)

In [332]:
arrhythmia_NEW_df['class'].value_counts()

11    245
1     205
Name: class, dtype: int64

### STEP 3: Replace Regular Heartbeat with '0' as a permanent "probability" value 

In [333]:
for n in arrhythmia_NEW_df['class']:
    if n == 11:
        arrhythmia_NEW_df = arrhythmia_NEW_df.replace([n], 0)

In [334]:
arrhythmia_NEW_df['class'].value_counts()

0    245
1    205
Name: class, dtype: int64

## 11) Converting Target('Class') to "Int64" 

### (NOT NEEDED BECAUSE STEPS 7-8 NOT RAN)

In [335]:
arrhythmia_NEW_df['class'].dtype

dtype('int64')

In [203]:
#arrhythmia_NEW_df['class'] = arrhythmia_NEW_df['class'].astype("int64")

In [204]:
#arrhythmia_NEW_df['class'].dtype

dtype('int64')

In [205]:
#arrhythmia_NEW_df['class'].value_counts()

0    245
1    205
Name: class, dtype: int64

In [336]:
type(arrhythmia_NEW_df)

pandas.core.frame.DataFrame

## 12) Splitting Data into Train and Test Sets...

In [371]:
X = arrhythmia_NEW_df.drop('class', axis=1).values
y = arrhythmia_NEW_df['class'].values 

# Split into training and test set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state=42, stratify=y)

#Standardize
sc= StandardScaler()
X_train=sc.fit_transform(X_train)
X_test=sc.fit_transform(X_test)

In [338]:
X_train.shape

(315, 194)

In [339]:
X_test.shape

(135, 194)

In [340]:
y_test.shape

(135,)

In [341]:
y_train.shape

(315,)

In [342]:
arrhythmia_NEW_df

age  sex  height  weight  QRSduration  PRinterval  Q-Tinterval  \
0     75    0     190      80           91         193          371   
1     56    0     165      64           81         174          401   
2     54    0     172      95          138         163          386   
3     55    0     175      94          100         202          380   
4     75    0     190      80           88         181          360   
..   ...  ...     ...     ...          ...         ...          ...   
447   53    0     160      70           80         199          382   
448   37    0     190      85          100         137          361   
449   36    0     166      68          108         176          365   
450   32    0     155      55           93         106          386   
451   78    0     160      70           79         127          364   

     Tinterval  Pinterval  QRS  ...  chV4_PwaveAmp  chV4_QRSA  chV4_QRSTA  \
0          174        121  -16  ...            0.9       17.7        70.7   
1          149         39   25  ...            0.5       11.8        34.6   
2          185        102   96  ...            0.5       -3.0        20.7   
3          179        143   28  ...            0.1       28.8        63.1   
4          177        103  -16  ...           -0.1       16.2        63.2   
..         ...        ...  ...  ...            ...        ...         ...   
447        154        117  -37  ...            0.8      -20.1        -9.5   
448        201         73   86  ...            0.7       69.2       129.3   
449        194        116  -85  ...            0.0      -71.2      -161.4   
450        218         63   54  ...            0.6       17.5        56.2   
451        138         78   28  ...            0.5       44.0        60.8   

     chV5_JJwaveAmp  chV5_RwaveAmp  chV5_SwaveAmp  chV5_SPwaveAmp  chV5_QRSA  \
0              -0.4           13.5           -4.0             0.0       25.5   
1              -0.4            0.0           -2.4             0.0       21.6   
2               1.3           11.1           -3.4             0.0       11.5   
3               0.1           15.2           -3.7             0.0       36.8   
4              -0.2            9.1           -0.9             0.0       21.7   
..              ...            ...            ...             ...        ...   
447             0.1            4.1           -8.3             0.0       -8.4   
448            -0.7           21.2           -2.8             0.0       50.7   
449             0.2           22.0          -30.8             0.0      -39.6   
450             0.1           15.3           -3.5             0.0       29.7   
451            -0.4           12.5           -2.9             0.0       26.7   

     chV6_SPwaveAmp  class  
0               0.0      1  
1               0.0      1  
2               0.0      1  
3               0.0      0  
4               0.0      1  
..              ...    ...  
447             0.0      0  
448             0.0      1  
449             0.0      1  
450             0.0      0  
451             0.0      0  

[450 rows x 195 columns]

### Tensorflow estimator ValueError: logits and labels must have the same shape ((?, 1) vs (?,))

### You should reshape your labels as 2d-tensor (the first dimension will be the batch dimension and the second the scalar label):

In [237]:
#y_train = np.asarray(y_train).astype('float32').reshape((-1,1))
#y_test = np.asarray(y_test).astype('float32').reshape((-1,1))

# NEURAL NETWORK (USING KERAS)

The Keras workflow has 4 steps. 

1) First, you specify the architecture,
        
        - How many layers do you want? 
        - How many nodes in each layer? 
        - What activation function do you want to use in each layer? 

2) Compile the model. 
    
        - This specifies the loss function, 
        - Some details about how optimization works. 

3) Then you fit the model. 

        - Which is that cycle of back-propagation and optimization of model weights with your data. 
        
4) And finally you will want to use your model to make predictions. 


After you've specified a model, the next task is to compile it, which sets up the network for optimization, for instance creating an internal function to do back-propagation efficiently. The compile methods

2. Why you need to compile your model
has two important arguments for you to choose. The first is what optimizer to use, which controls the learning rate.

The second thing you specify is the loss function.

## Specify/Create the Neural Network/Model

In [343]:
from keras.layers import Dense
from keras.models import Sequential

In [344]:
#allows you to seperate the binary outcomes 
from keras.utils.np_utils import to_categorical

y_train = to_categorical(y_train, 2)
y_test = to_categorical(y_test, 2)

## USE to_categorical (above)

In [345]:
n_cols = X_train.shape[1] #194 columns excluding target/class
n_cols

194

In [346]:
# Set up the model: model
model = Sequential()

# Add the first layer
model.add(Dense(100, activation='relu', input_shape = (n_cols,)))

# Add the second layer
#model.add(Dense(100, activation='relu'))

# Add the third layer
model.add(Dense(100, activation='relu'))

# Add the output layer
model.add(Dense(2, activation='sigmoid')) #2 nodes for the 2 possible outcomes --irregular vs regular

#softmax

## Compile the Model

In [347]:
from tensorflow import keras
from tensorflow.keras import layers

In [348]:
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=["accuracy"])

## Fit the Model

In [349]:
model.fit(X_train, y_train, epochs = 100)

Epoch 1/100
10/10 [==============================] - 2s 9ms/step - loss: 0.7305 - accuracy: 0.5429
Epoch 2/100
10/10 [==============================] - 0s 5ms/step - loss: 0.6011 - accuracy: 0.6571
Epoch 3/100
10/10 [==============================] - 0s 5ms/step - loss: 0.5234 - accuracy: 0.7841
Epoch 4/100
10/10 [==============================] - 0s 4ms/step - loss: 0.4631 - accuracy: 0.8254
Epoch 5/100
10/10 [==============================] - 0s 5ms/step - loss: 0.4142 - accuracy: 0.8381
Epoch 6/100
10/10 [==============================] - 0s 8ms/step - loss: 0.3685 - accuracy: 0.8476
Epoch 7/100
10/10 [==============================] - 0s 6ms/step - loss: 0.3282 - accuracy: 0.8889
Epoch 8/100
10/10 [==============================] - 0s 6ms/step - loss: 0.2928 - accuracy: 0.9016
Epoch 9/100
10/10 [==============================] - 0s 4ms/step - loss: 0.2591 - accuracy: 0.9206
Epoch 10/100
10/10 [==============================] - 0s 3ms/step - loss: 0.2307 - accuracy: 0.9333
Epoch 11/

10/10 [==============================] - 0s 6ms/step - loss: 4.7414e-04 - accuracy: 1.0000
Epoch 83/100
10/10 [==============================] - 0s 8ms/step - loss: 4.6058e-04 - accuracy: 1.0000
Epoch 84/100
10/10 [==============================] - 0s 4ms/step - loss: 4.4780e-04 - accuracy: 1.0000
Epoch 85/100
10/10 [==============================] - 0s 9ms/step - loss: 4.3460e-04 - accuracy: 1.0000
Epoch 86/100
10/10 [==============================] - 0s 4ms/step - loss: 4.2412e-04 - accuracy: 1.0000
Epoch 87/100
10/10 [==============================] - 0s 4ms/step - loss: 4.1135e-04 - accuracy: 1.0000
Epoch 88/100
10/10 [==============================] - 0s 8ms/step - loss: 4.0103e-04 - accuracy: 1.0000
Epoch 89/100
10/10 [==============================] - 0s 7ms/step - loss: 3.8963e-04 - accuracy: 1.0000
Epoch 90/100
10/10 [==============================] - 0s 11ms/step - loss: 3.7915e-04 - accuracy: 1.0000
Epoch 91/100
10/10 [==============================] - 0s 10ms/step - loss: 3

## Make Predictions

In [351]:
# ERROR: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

# Calculate predictions: predictions
predictions = model.predict(X_test)

In [352]:
predictions

array([[9.96794581e-01, 4.24540043e-03],
       [9.75700459e-06, 9.99954939e-01],
       [9.98951077e-01, 9.36597586e-04],
       [9.99989092e-01, 1.32760561e-05],
       [8.88536513e-01, 1.24453664e-01],
       [9.98091102e-01, 1.94752216e-03],
       [6.71067357e-01, 3.83071601e-01],
       [3.13669443e-04, 9.99586225e-01],
       [9.99986887e-01, 1.61183380e-05],
       [1.47351220e-09, 1.00000000e+00],
       [8.41066957e-01, 1.18926048e-01],
       [1.16725445e-01, 8.93856049e-01],
       [5.56567311e-03, 9.93367195e-01],
       [5.06111979e-03, 9.94741261e-01],
       [9.99930024e-01, 8.65243783e-05],
       [9.67145920e-01, 5.41410446e-02],
       [1.14029646e-03, 9.98560309e-01],
       [8.17699103e-11, 1.00000000e+00],
       [3.97538861e-05, 9.99845028e-01],
       [9.98864591e-01, 1.22669339e-03],
       [9.99960303e-01, 5.01835093e-05],
       [1.17203026e-05, 9.99969840e-01],
       [6.73095499e-08, 1.00000000e+00],
       [1.10482192e-18, 1.00000000e+00],
       [8.542988

## Calculate Accuracy and look at Classification Report

In [354]:
# Evaluate accuracy
#calculate accuracy
from sklearn.metrics import balanced_accuracy_score as accuracy

accuracy_of_model = accuracy(y_test, predictions)
print(accuracy_of_model)

#model.evalute(x_test, y_test)

ValueError: Classification metrics can't handle a mix of multilabel-indicator and continuous-multioutput targets

In [355]:
print(classification_report(y_test, predictions))

ValueError: Classification metrics can't handle a mix of multilabel-indicator and continuous-multioutput targets

#### 4.	Write another algorithm to predict the same result as the previous question using either KNN or logistic regression.

In [358]:
from sklearn.linear_model import LogisticRegression

In [359]:
#Logistic regression approach
regression = LogisticRegression(random_state=42).fit(X_train, y_train)

#regression.fit(X_train, y_train)

y_predicted = regression.predict(X_test)

In [360]:
regression.score(X_test, y_test)

accuracy = regression.score(X_test, y_test)
print("accuracy = ", accuracy * 100, "%")

accuracy =  74.07407407407408 %


In [361]:
print(classification_report(y_test, y_predicted))

              precision    recall  f1-score   support

           0       0.76      0.77      0.77        74
           1       0.72      0.70      0.71        61

    accuracy                           0.74       135
   macro avg       0.74      0.74      0.74       135
weighted avg       0.74      0.74      0.74       135



#### 5.	Create a neural network using pytorch to predict the same result as question 3. 

In [362]:
import torch

In [372]:
# import torch neural network
import torch.nn as nn

# the following is imported because it has activation functions in it which is  convert the node input to 
#an ouput so that the algorithm can learn the complex patterns in the data.
import torch.nn.functional as F

#create tensors from our data = matrices (select tensor type that we need)

#redefine X_train and X_test based on the datatype we need to use(float)--pass in the matrices above(X_train, X_test)
X_train = torch.FloatTensor(X_train) 
X_test = torch.FloatTensor(X_test)

In [373]:
#redefine y_train and y_test based on the datatype we need to use(int-use long)--pass in the matrices above(y_train, y_test)
y_train = torch.LongTensor(y_train)

In [374]:
y_test = torch.LongTensor(y_test)

#y_test = torch.LongTensor(y_test.to_numpy())

In [375]:
#create an artificial neural network using class that accepts a NN module
class ANN_Model(nn.Module):
    
    #initiialize the class and its attributes that everything you create from that class will have
    # self: refers to the object itself
    # the other parameters (data we pass in) are part of the NN module we imported
        # input_features-->8 columns excluding target variable
        # hidden layers--> 20 nodes or perceptrons
        # 2 output features--> regular, irregular
    def __init__(self, input_features=194,hidden1=40,hidden2=40,out_features=2):
        
        super().__init__() #super is a computed indirect reference. So, it isolates changes
        # and makes sure that children in the layers of multiple inheritence are calling
        #the right parents, "parent and child need to be able to talk to each other" allows the layers and 
        #hence the nodes talk to each other. this is how we keep track of the trail
        
        #first connection is referring to all the connections between the input features 
        #and the first layer--hidden1.
        self.layer_1_connection = nn.Linear(input_features, hidden1)
        
        #second connection is referring to all the connections between hidden1 
        #and hidden2.
        self.layer_2_connection = nn.Linear(hidden1, hidden2)
        
        #third connection is referring to all the connections between hidden2 
        #and the output.
        self.out = nn.Linear(hidden2, out_features)
    
    #class methods/functions-->build in forward propagation into the class   
    def forward(self, x):
        #apply activation functions-->relu (from torch.nn.functional as F)
        x = F.relu(self.layer_1_connection(x))
        x = F.relu(self.layer_2_connection(x))
        x = self.out(x)
        return x

In [376]:
torch.manual_seed(42)

#create instance of model
    # by making the variable 'ann' equal to the class, it is automatically passing in the NN created 
    # within the class above
ann = ANN_Model()

In [377]:
#loss function
loss_function = nn.CrossEntropyLoss()

#optimizer--a way to improve the model on top of applying activation function, loss functions 
# pass in a learning rate-- you wouldnt want it to learn too quickly cuz that would make haste-y decisions
optimizer = torch.optim.Adam(ann.parameters(),lr=0.02)

In [378]:
#run model through multiple epochs/iterations
#this is how we will record the loss

final_loss = []
n_epochs = 500
for epoch in range(n_epochs):
    y_pred = ann.forward(X_train) #ann, name of model, forward propagate the training data
    loss = loss_function(y_pred, y_train)
    final_loss.append(loss)
    
    if epoch % 10 == 1: #for every 10 iterations...
        print(f'Epoch number: {epoch} with loss: {loss}')
    
    #Impliment optimizer with gradient decent
    # .zero_grad() --> allows us to work on just imporving our loss function using backward 
    #propagation w/o having forward rpopagation interfere 
    optimizer.zero_grad() #zero the gradient before running backwards propagation so we are starting fresh
    loss.backward() 
    optimizer.step() #perform one optimization step each epoch

#as we can see, there is lower loss as we increase the iterations



Epoch number: 1 with loss: 0.5679716467857361
Epoch number: 11 with loss: 0.07899633795022964
Epoch number: 21 with loss: 0.00504554808139801
Epoch number: 31 with loss: 6.175383896334097e-05
Epoch number: 41 with loss: 7.432846814481309e-06
Epoch number: 51 with loss: 3.114694436590071e-06
Epoch number: 61 with loss: 1.2545028766908217e-06
Epoch number: 71 with loss: 7.307621103791462e-07
Epoch number: 81 with loss: 5.438164407678414e-07
Epoch number: 91 with loss: 4.4655860165221384e-07
Epoch number: 101 with loss: 3.8411658920267655e-07
Epoch number: 111 with loss: 3.390824758753297e-07
Epoch number: 121 with loss: 3.054014712233766e-07
Epoch number: 131 with loss: 2.792892246361589e-07
Epoch number: 141 with loss: 2.5582602347640204e-07
Epoch number: 151 with loss: 2.3690404304943513e-07
Epoch number: 161 with loss: 2.2214491934846592e-07
Epoch number: 171 with loss: 2.077642022868531e-07
Epoch number: 181 with loss: 1.960325448635558e-07
Epoch number: 191 with loss: 1.854362494668

In [379]:
#store predictions in list
y_pred = []

with torch.no_grad(): #start with zero gradiation and iterate thru the below for loop...
    for i, data in enumerate(X_test): # i --> keep track of index, enumerate thru test set
        prediction = ann(data)
        y_pred.append(prediction.argmax()) #look up argmax (Alexis did not have explanation)

In [380]:
from sklearn.metrics import classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.76      0.82      0.79        74
           1       0.76      0.69      0.72        61

    accuracy                           0.76       135
   macro avg       0.76      0.76      0.76       135
weighted avg       0.76      0.76      0.76       135



#### 6.	Compare the performance of the neural networks to the other model you created. Which performed better? Why do you think that is?

# TROUBLESHOOTING STEPS:

# DONE: NEXT STEP--> .values

# DONE: SKIP PREPROCESSING STEPS 7-8

# DONE: (just keep standard scalar)

# NEXT STEP--> GO BACK TO ONE HOT ENCODING
